### PR TITLE
feat: add x/fibre keeper, module, and CLI

### PR DESCRIPTION
**Pull request by rootulp**
## Summary

Add the on-chain fibre module implementation (PR 1 of the fibre port from celestia-app-fibre).

### New files

**Keeper** (`x/fibre/keeper/`):
- `keeper.go` — Core keeper with escrow account management, withdrawal processing, processed payment tracking, and payment promise validation (stateless + stateful)
- `msg_server.go` — MsgServer handling DepositToEscrow, RequestWithdrawal, PayForFibre, PaymentPromiseTimeout, UpdateFibreParams
- `grpc_query.go` — QueryServer for params, escrow accounts, withdrawals, payment status, and payment promise validation
- `abci.go` — BeginBlocker processing available withdrawals and pruning old processed payments
- `genesis.go` — InitGenesis/ExportGenesis with iterate helpers

**Module** (`x/fibre/`):
- `module.go` — Full AppModule implementation (AppModuleBasic, HasGenesis, HasServices, BeginBlock)

**CLI** (`x/fibre/client/cli/`):
- `query.go` — Query commands: params, escrow-account, withdrawals, is-payment-processed
- `tx.go` — Tx commands: deposit-to-escrow, request-withdrawal, pay-for-fibre, payment-promise-timeout

**Fibre stubs** (`fibre/`):
- `commitment.go` — Minimal Commitment type ([32]byte) needed by keeper
- `payment_promise.go` — PaymentPromise type with validation, signing, and proto conversion
- `validator/set.go` — Validator Set wrapper with GetByAddress
- `validator/signature_set.go` — SignatureSet for collecting validator signatures with voting power thresholds

**Tests** (`x/fibre/keeper/`):
- `keeper_test.go` — Keeper unit tests (params, escrow, withdrawals, processed payments, validation)
- `msg_server_test.go` — MsgServer tests for all 5 message types with validator signature generation
- `abci_test.go` — BeginBlocker tests (withdrawal processing, pruning, edge cases)
- `mocks_test.go` — MockBankKeeper and MockStakingKeeper

### Notes

- Depends on #6703 (x/fibre/types/msgs_test.go)
- The `fibre/` stubs are minimal — just enough to satisfy keeper imports. They will be extended in a follow-up PR (fibre networking package)
- Module is not yet wired into `app/app.go` — that will be a separate PR (app wiring)

## Test plan

- [x] `go test ./x/fibre/keeper/...` passes
- [x] `go test ./x/fibre/types/...` passes
- [x] `make build-standalone` succeeds
- [x] `golangci-lint` passes with 0 issues

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: fibre/validator/set.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+)
+
+// Set wraps [*core.ValidatorSet] with the height at which it is valid.
+type Set struct {
+	*core.ValidatorSet
+	Height uint64
+}
+
+// GetByAddress finds a validator in the set by address.
+// Returns the validator pointer from ValidatorSet.Validators (required for shard map lookups)
```
Comment (rach-id) [inline fibre/validator/set.go:15]: what does `(required for shard map lookups)` mean here?

File: fibre/validator/signature_set.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	cmtmath "github.com/cometbft/cometbft/libs/math"
+	core "github.com/cometbft/cometbft/types"
+)
+
+// SignatureSet collects and validates signatures from validators.
+// It is safe for concurrent use.
+// Signatures are returned in validator set order by [Signatures],
+// with nil entries for validators that did not sign.
+type SignatureSet struct {
+	requiredBytesSigned    []byte
```
Comment (rach-id) [inline fibre/validator/signature_set.go:17]: could use an explanation on what this is

File: fibre/payment_promise.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+	pbMsg := &types.PaymentPromise{}
+	if err := gogoproto.Unmarshal(data, pbMsg); err != nil {
+		return err
+	}
+
+	return p.FromProto(pbMsg)
+}
+
+// FromProto converts the [PaymentPromise] from its protobuf representation.
+func (p *PaymentPromise) FromProto(pbMsg *types.PaymentPromise) error {
```
Comment (rach-id) [inline fibre/payment_promise.go:74]: would be easier for this to create the PaymentPromise inside the function and return instead of expecting an empty one

File: x/fibre/client/cli/tx.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+	return cmd
+}
+
+// CmdPaymentPromiseTimeout implements the payment-promise-timeout transaction command.
+func CmdPaymentPromiseTimeout() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "payment-promise-timeout [payment-promise-json]",
+		Args:  cobra.ExactArgs(1),
+		Short: "Process a timed-out payment promise",
```
Comment (rach-id) [inline x/fibre/client/cli/tx.go:182]: what's a timed-out payment process?

File: fibre/payment_promise.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+		Signature:         p.Signature,
+	}
+}
+
+// Validate performs stateless validation on the [PaymentPromise].
+// It verifies all field constraints and validates the [PaymentPromise.Signature] using [PaymentPromise.SignerKey].
+func (p *PaymentPromise) Validate() error {
+	// signer key must be valid secp256k1 public key (33 bytes)
+	if p.SignerKey == nil || len(p.SignerKey.Key) != secp256k1.PubKeySize {
+		return fmt.Errorf("signer key must be %d bytes, got %d", secp256k1.PubKeySize, len(p.SignerKey.Key))
```
Comment (rach-id) [inline fibre/payment_promise.go:124]: [blocking]
if `p.SignerKey == nil`, `p.SignerKey.Key` would panic. 

By Claude

```suggestion
	if p.SignerKey == nil || len(p.SignerKey.Key) != secp256k1.PubKeySize {
		return fmt.Errorf("signer key must be %d bytes, got %d", secp256k1.PubKeySize, len(p.SignerKey))
```

File: fibre/validator/signature_set.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	CollectedPower int64
+	RequiredPower  int64
+}
+
+func (e *NotEnoughSignaturesError) Error() string {
+	switch {
+	case e.CollectedPower < e.RequiredPower:
+		return fmt.Sprintf("not enough voting power: collected %d, required %d", e.CollectedPower, e.RequiredPower)
+	default:
+		panic("unreachable")
```
Comment (rach-id) [inline fibre/validator/signature_set.go:98]: this unnecessarily adds a case that panics. It's useful to block dev mistakes, but having it in production would halt the Fibre server. We should just return a different error IMO

File: fibre/validator/signature_set.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		minRequiredVotingPower: minRequiredVotingPower,
+		validators:             s.Validators,
+		signatures:             make(map[string][]byte, s.Size()),
+	}
+}
+
+// Add validates and adds a signature from the given validator.
+// Returns an error if the signature is invalid.
+// Returns true if enough signatures have been collected to meet both thresholds.
+func (ss *SignatureSet) Add(val *core.Validator, signature []byte) (bool, error) {
```
Comment (rach-id) [inline fibre/validator/signature_set.go:41]: [nit]
to avoid mistakes, we could check whether this signature was already added. If so, return an error or just nil. Otherwise, we risk increasing the voting power by mistake if we call this method twice on the same validator/signature

File: x/fibre/keeper/msg_server.go [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+	}
+
+	return &types.MsgUpdateFibreParamsResponse{}, nil
+}
+
+// calculatePaymentAmount calculates the payment amount for a fibre blob based on its size and gas parameters
+func (ms msgServer) calculatePaymentAmount(ctx sdk.Context, blobSize uint32) sdk.Coin {
+	params := ms.GetParams(ctx)
+	// TODO: this assumes 1 utia per gas which may not be correct.
+	return sdk.NewInt64Coin(appconsts.BondDenom, int64(blobSize*params.GasPerBlobByte))
```
Comment (rach-id) [inline x/fibre/keeper/msg_server.go:301]: Claude:
```
int64(blobSize * params.GasPerBlobByte)

  blobSize is uint32 and GasPerBlobByte is uint32. Their product can overflow uint32 (max ~4.2B). The multiplication happens in uint32 space before the int64 cast. Should cast first:

  int64(uint64(blobSize) * uint64(params.GasPerBlobByte))
```

File: x/fibre/keeper/msg_server.go [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+	event := types.NewEventPaymentPromiseTimeout(msg.Signer, escrowSigner, promiseHash)
+	if err := ctx.EventManager().EmitTypedEvent(event); err != nil {
+		return nil, err
+	}
+
+	return &types.MsgPaymentPromiseTimeoutResponse{}, nil
+}
+
+// UpdateFibreParams updates the fibre module parameters
+func (ms msgServer) UpdateFibreParams(goCtx context.Context, msg *types.MsgUpdateFibreParams) (*types.MsgUpdateFibreParamsResponse, error) {
```
Comment (rach-id) [inline x/fibre/keeper/msg_server.go:277]: Claude:
```
Missing UpdateFibreParams validation — msg_server.go:3226

  UpdateFibreParams sets params without calling params.Validate(). Governance could set GasPerBlobByte = 0, WithdrawalDelay to negative duration, etc. Should validate before setting
```

File: fibre/validator/set.go (hunk 1 above)
Comment (rootulp) [reply fibre/validator/set.go:15]: Will try to answer your questions but this is copied from celestia-app-fibre so @Wondertan is the expert here.

I infer that when clients look up key vals in the shard map, they then call GetByAddress on the validator.

Ref: https://github.com/celestiaorg/celestia-app-fibre/blob/5346c56ef040c4d819da603381cf751227ee0b82/fibre/validator/set.go#L20

File: fibre/validator/signature_set.go (hunk 2 above)
Comment (rootulp) [reply fibre/validator/signature_set.go:17]: Copied from celestia-app-fibre. Can create a follow-up issue to add godocs to not block the celestia-app-fibre -> celestia-app migration.

Ref: https://github.com/celestiaorg/celestia-app-fibre/blob/5346c56ef040c4d819da603381cf751227ee0b82/fibre/validator/signature_set.go#L17

File: fibre/payment_promise.go (hunk 3 above)
Comment (rootulp) [reply fibre/payment_promise.go:74]: Copied from celestia-app-fibre. Can create a follow-up issue to not block the celestia-app-fibre -> celestia-app migration.

Ref: https://github.com/celestiaorg/celestia-app-fibre/blob/5346c56ef040c4d819da603381cf751227ee0b82/fibre/payment_promise.go#L73-L74

File: fibre/validator/signature_set.go (hunk 2 above)
Comment (rootulp) [reply fibre/validator/signature_set.go:17]: https://github.com/celestiaorg/celestia-app/issues/6711

File: x/fibre/client/cli/tx.go (hunk 4 above)
Comment (rootulp) [reply x/fibre/client/cli/tx.go:182]: A payment promise that a user didn't submit on chain for some timeout period (I think currently configured to 24 hours).

If the payment promise didn't get submitted on-chain then validators are retaining the data but the user was never charged for the Fibre data. 

A timed out payment promise is the mechanism that a watchdog node uses to ensure that all payment promises that it sees are eventually submit on-chain in order to prevent users from getting DA for free. A timed out payment promise gets included on-chain and the state machine deducts fees from the user.

File: fibre/payment_promise.go (hunk 5 above)
Comment (rootulp) [reply fibre/payment_promise.go:124]: Good point. I'm not sure how this exists in celestia-app-fibre. cc: @Wondertan 

Ref: https://github.com/celestiaorg/celestia-app-fibre/blob/5346c56ef040c4d819da603381cf751227ee0b82/fibre/payment_promise.go#L122-L124

FLUP should be to add unit tests to confirm this error doesn't throw a panic.

Comment (rootulp) [reply fibre/payment_promise.go:124]: https://github.com/celestiaorg/celestia-app/issues/6712

File: fibre/validator/signature_set.go (hunk 6 above)
Comment (rootulp) [reply fibre/validator/signature_set.go:98]: I totally agree. Saw it in the code copy + paste but didn't want to modify it directly. Creating a FLUP to remove the `panic`.

Comment (rootulp) [reply fibre/validator/signature_set.go:98]: https://github.com/celestiaorg/celestia-app/issues/6713

File: fibre/validator/signature_set.go (hunk 7 above)
Comment (rootulp) [reply fibre/validator/signature_set.go:41]: Great idea: https://github.com/celestiaorg/celestia-app/issues/6714

File: x/fibre/keeper/msg_server.go (hunk 8 above)
Comment (rootulp) [reply x/fibre/keeper/msg_server.go:301]: https://github.com/celestiaorg/celestia-app/issues/6715

File: x/fibre/keeper/msg_server.go (hunk 9 above)
Comment (rootulp) [reply x/fibre/keeper/msg_server.go:277]: https://github.com/celestiaorg/celestia-app/issues/6716

File: x/fibre/keeper/msg_server.go (hunk 8 above)
Comment (rootulp) [reply x/fibre/keeper/msg_server.go:301]: https://github.com/celestiaorg/celestia-app/issues/6715

**Comment by rootulp** [comment]
@rach-id thanks for the thorough review. I agree with all of your feedback. I spun out follow-up issues because I'd like to merge this PR to unblock the rest of the code migration. Will address your follow-up issues in parallel.
